tests/system_tests/c++/test-output-files.cpp

```diff
@@ -0,0 +1,121 @@
+/*----------------------------------------------------------------------------*\
+ (c) Crown copyright 2025 Met Office. All rights reserved.
+ The file LICENCE, distributed with this code, contains details of the terms
+ under which the code may be used.
+\*----------------------------------------------------------------------------*/
+
+// System test to check that the file outputs are correct
+
+#include "vernier.h"
+#include "vernier_mpi.h"
+
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <string>
+
+/*
+ * Check the first few lines of the output file
+ */
+bool check_output_file(std::string path, std::string format) {
+  std::filebuf fb;
+  std::string buffer;
+  std::string expected;
+
+  if (!fb.open(path, std::ios::in)) {
+    std::cerr << "failed to open " << path << std::endl;
+    return false;
+  }
+
+  std::istream input(&fb);
+
+  // Check the header lines
+  std::getline(input, buffer);
+  if (buffer.compare("") != 0) {
+    std::cerr << "Invalid line: " << buffer << std::endl;
+    return false;
+  }
+
+  std::getline(input, buffer);
+  if (buffer.compare(0, 8, "MPI task") != 0) {
+    std::cerr << "Invalid line: " << buffer << std::endl;
+    return false;
+  }
+
+  // Match the next line to the start of an output format
+  if (format.compare("drhook") == 0) {
+    expected = "Profiling on ";
+  } else if (format.compare("threads") == 0) {
+    // Skip extra blank line in threads format
+    std::getline(input, buffer);
+    expected = "region_name@thread_id";
+  } else {
+    std::cerr << "unknown format" << std::endl;
+    return false;
+  }
+
+  std::getline(input, buffer);
+  if (buffer.compare(0, expected.length(), expected) != 0) {
+    std::cerr << "Invalid next line" << std::endl;
+    std::cerr << "Got:      " << buffer << std::endl;
+    std::cerr << "Expected: " << expected << std::endl;
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * Generate some profiler output and check the contents
+ */
+bool create_output(std::string mode, std::string format) {
+  std::string path;
+  meto::vernier.init(MPI_COMM_WORLD);
+
+  // Create some data
+  auto vnr_handle = meto::vernier.start("main");
+  meto::vernier.stop(vnr_handle);
+
+  setenv("VERNIER_OUTPUT_MODE", mode.c_str(), 1);
+  setenv("VERNIER_OUTPUT_FORMAT", format.c_str(), 1);
+
+  meto::vernier.write();
+
+  meto::vernier.finalize();
+
+  if (mode.compare("multi") == 0) {
+    path = "vernier-output-0";
+  } else if (mode.compare("single") == 0) {
+    path = "vernier-output-collated";
+  } else {
+    std::cerr << "unknown mode" << std::endl;
+    return false;
```

Comment:
I really don't like multiple return statements in a function, I think we'd be better off using a `GOTO` statement :stuck_out_tongue_winking_eye:

Reply:
It's a test, we have no logger to error out so return is fine.